File: parser/build_unified.py
"""Build unified sales mart from 1C operations and Bitrix deals."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

import pandas as pd

from parser.load_1c import resolve_agent
from parser.load_bitrix import resolve_agent_bitrix

UNIFIED_COLUMNS = [
    "source",
    "date",
    "agent_key",
    "agent_display",
    "agent_team",
    "agent_is_active",
    "agent_raw",
    "client",
    "client_id",
    "partner_or_supplier",
    "category",
    "channel",
    "card_type",
    "client_type",
    "request_type",
    "sales_amount",
    "profit_ex_vat",
    "service_fee",
    "is_refund",
    "raw_id",
    "date_fallback_used",
]


def _is_na(value: Any) -> bool:
    return value is None or value is pd.NA or (isinstance(value, float) and pd.isna(value))


def _as_str(value: Any) -> str | None:
    if _is_na(value):
        return None
    text = str(value).replace("\xa0", " ").strip()
    return text if text else None


def _as_date(value: Any) -> date | None:
    if _is_na(value):
        return None
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    return None


def _as_float(value: Any) -> float | None:
    if _is_na(value):
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _team_from_department(department: Any, department_map: dict[str, str]) -> str:
    if _is_na(department):
        return "Без команды"
    dept = str(department).replace("\xa0", " ").strip()
    if not dept:
        return "Без команды"
    for key, display in department_map.items():
        if display == dept:
            return key
    if dept in department_map:
        return dept
    return dept


def _pick_category(related_service_type: Any, category: Any) -> str | None:
    for value in (related_service_type, category):
        text = _as_str(value)
        if text:
            return text
    return None


def _apply_agent_fields(
    row: dict[str, Any],
    resolved: dict[str, Any],
    *,
    department: Any = None,
    department_map: dict[str, str] | None = None,
) -> None:
    row["agent_key"] = resolved["agent_key"]
    row["agent_display"] = resolved["name_display"] or resolved.get("agent_key")
    row["agent_is_active"] = resolved.get("is_active")

    team = resolved.get("team")
    if str(resolved["agent_key"]).startswith("unknown:") and department_map is not None:
        team = _team_from_department(department, department_map)
    row["agent_team"] = team if team else "Без команды"


def build_sales_unified(
    operations_1c: pd.DataFrame,
    deals_bitrix: pd.DataFrame,
    settings: dict[str, Any],
) -> pd.DataFrame:
    """Combine 1C (all rows) and Bitrix (success deals only) into sales_unified."""
    metrics = settings.get("metrics", {})
    success_value = metrics.get("bitrix_success_value", "Успех")
    department_map = settings.get("department_map", {})

    rows: list[dict[str, Any]] = []

    for _, record in operations_1c.iterrows():
        sales_amount = _as_float(record.get("sales_amount"))
        resolved = resolve_agent(
            record.get("agent"),
            settings,
            department=record.get("department"),
        )
        unified: dict[str, Any] = {
            "source": "1c",
            "date": _as_date(record.get("date_operation")),
            "agent_raw": _as_str(record.get("agent")),
            "client": _as_str(record.get("client")),
            "client_id": _as_str(record.get("id_crm")),
            "partner_or_supplier": _as_str(record.get("supplier")),
            "category": _pick_category(
                record.get("related_service_type"),
                record.get("category"),
            ),
            "channel": _as_str(record.get("channel")),
            "card_type": _as_str(record.get("card_type")),
            "client_type": None,
            "request_type": None,
            "sales_amount": sales_amount,
            "profit_ex_vat": _as_float(record.get("profit_ex_vat")),
            "service_fee": _as_float(record.get("service_fee")),
            "is_refund": bool(sales_amount is not None and sales_amount < 0),
            "raw_id": _as_str(record.get("order_no")),
            "date_fallback_used": None,
        }
        _apply_agent_fields(
            unified,
            resolved,
            department=record.get("department"),
            department_map=department_map,
        )
        rows.append(unified)

    bitrix_sales = deals_bitrix
    if "deal_result" in deals_bitrix.columns:
        bitrix_sales = deals_bitrix[deals_bitrix["deal_result"] == success_value]

    for _, record in bitrix_sales.iterrows():
        sales_amount = _as_float(record.get("sales_amount"))
        resolved = resolve_agent_bitrix(record.get("responsible_person"), settings)
        deal_no = record.get("deal_no")
        raw_id = None if _is_na(deal_no) else str(int(deal_no)) if float(deal_no).is_integer() else str(deal_no)

        unified = {
            "source": "bitrix",
            "date": _as_date(record.get("date_for_sales")),
            "agent_raw": _as_str(record.get("responsible_person")),
            "client": _as_str(record.get("client")),
            "client_id": _as_str(record.get("id_client")),
            "partner_or_supplier": _as_str(record.get("partner")),
            "category": _as_str(record.get("category")),
            "channel": _as_str(record.get("channel")),
            "card_type": _as_str(record.get("card_type")),
            "client_type": _as_str(record.get("client_type")),
            "request_type": _as_str(record.get("request_type")),
            "sales_amount": sales_amount,
            "profit_ex_vat": _as_float(record.get("profit_ex_vat")),
            "service_fee": _as_float(record.get("service_fee")),
            "is_refund": False,
            "raw_id": raw_id,
            "date_fallback_used": bool(record.get("date_fallback_used"))
            if not _is_na(record.get("date_fallback_used"))
            else None,
        }
        _apply_agent_fields(unified, resolved)
        rows.append(unified)

    if not rows:
        return pd.DataFrame(columns=UNIFIED_COLUMNS)

    frame = pd.DataFrame(rows)
    for column in UNIFIED_COLUMNS:
        if column not in frame.columns:
            frame[column] = None
    return frame[UNIFIED_COLUMNS]


def build_agents_dim(
    settings: dict[str, Any],
    sales_unified: pd.DataFrame,
) -> pd.DataFrame:
    """Settings agents plus unknown agent statistics from unified sales."""
    rows: list[dict[str, Any]] = []

    for agent in settings.get("agents", []):
        rows.append(
            {
                "agent_key": agent.get("agent_key"),
                "name_display": agent.get("name_display"),
                "names_1c": "|".join(agent.get("names_1c", [])),
                "names_bitrix": "|".join(agent.get("names_bitrix", [])),
                "team": agent.get("team"),
                "is_active": agent.get("is_active"),
                "is_unknown": False,
                "count_1c": 0,
                "count_bitrix": 0,
                "count_total": 0,
            }
        )

    known_keys = {row["agent_key"] for row in rows}
    if not sales_unified.empty and "agent_key" in sales_unified.columns:
        unknown = sales_unified[sales_unified["agent_key"].astype(str).str.startswith("unknown:")]
        if not unknown.empty:
            grouped = (
                unknown.groupby(["agent_key", "agent_display", "agent_team", "source"], dropna=False)
                .size()
                .reset_index(name="count")
            )
            stats: dict[str, dict[str, Any]] = {}
            for _, item in grouped.iterrows():
                key = item["agent_key"]
                stats.setdefault(
                    key,
                    {
                        "agent_key": key,
                        "name_display": item["agent_display"],
                        "names_1c": "",
                        "names_bitrix": "",
                        "team": item["agent_team"],
                        "is_active": None,
                        "is_unknown": True,
                        "count_1c": 0,
                        "count_bitrix": 0,
                        "count_total": 0,
                    },
                )
                if item["source"] == "1c":
                    stats[key]["count_1c"] += int(item["count"])
                elif item["source"] == "bitrix":
                    stats[key]["count_bitrix"] += int(item["count"])
                stats[key]["count_total"] += int(item["count"])

            for key, item in stats.items():
                if key not in known_keys:
                    rows.append(item)

    return pd.DataFrame(rows)

File: lib/build_unified.php
<?php
/**
 * build_unified.php
 *
 * Собирает единую таблицу sales_unified:
 * - ВСЕ строки 1С (в т.ч. возвраты с отрицательной sales_amount)
 * - сделки Битрикс с результатом «Успех»
 * - сделки Битрикс на стадиях возврата — с отрицательной sales_amount
 *
 * Источники НЕ склеиваем между собой (нет дедупа 1С↔Битрикс).
 * Прибыль везде = profit_ex_vat.
 */

declare(strict_types=1);

require_once __DIR__ . '/settings.php';

/** Стадии CRM, которые считаем возвратами (сумма в unified идёт со знаком минус). */
function bitrix_refund_statuses(?array $settings = null): array
{
    $settings = $settings ?? [];
    $configured = $settings['metrics']['bitrix_refund_statuses'] ?? null;
    if (is_array($configured) && $configured !== []) {
        return array_values(array_filter(array_map('clean_str', $configured)));
    }
    return ['Возврат подтверждён', 'Возврат'];
}

function is_bitrix_refund_deal(array $record, ?array $settings = null): bool
{
    $status = clean_str($record['deal_status'] ?? null);
    if ($status === null) {
        return false;
    }
    foreach (bitrix_refund_statuses($settings) as $refundStatus) {
        if ($status === $refundStatus) {
            return true;
        }
    }
    return mb_stripos($status, 'возврат') !== false;
}

/** Синонимы категорий: «Отельный билет» → «Отель» и т.п. */
function normalize_category(?string $category, ?array $settings = null): ?string
{
    $category = clean_str($category);
    if ($category === null) {
        return null;
    }
    $map = $settings['category_aliases'] ?? [];
    if (!is_array($map) || $map === []) {
        return $category;
    }
    $canonical = $map[$category] ?? null;
    return clean_str(is_string($canonical) ? $canonical : null) ?? $category;
}

/** Регистронезависимый lookup в map alias → canonical. */
function alias_lookup(?string $value, array $map): ?string
{
    $value = clean_str($value);
    if ($value === null || $map === []) {
        return null;
    }
    if (isset($map[$value]) && is_string($map[$value])) {
        return clean_str($map[$value]);
    }
    $lower = mb_strtolower($value);
    foreach ($map as $from => $to) {
        if (!is_string($from) || !is_string($to)) {
            continue;
        }
        if (mb_strtolower($from) === $lower) {
            return clean_str($to);
        }
    }
    return null;
}

/**
 * Тип запроса: явный request_type, иначе вывод из категории 1С (Air Tickets → Travel).
 *
 * @param list<string|null> $categoryHints сырые category / related_service_type / нормализованная category
 */
function resolve_request_type(?string $requestType, array $categoryHints, ?array $settings = null): ?string
{
    $settings = $settings ?? [];
    $rtMap = is_array($settings['request_type_aliases'] ?? null) ? $settings['request_type_aliases'] : [];
    $catMap = is_array($settings['category_to_request_type'] ?? null) ? $settings['category_to_request_type'] : [];

    $requestType = clean_str($requestType);
    if ($requestType !== null) {
        return alias_lookup($requestType, $rtMap) ?? $requestType;
    }
    foreach ($categoryHints as $hint) {
        $mapped = alias_lookup(is_string($hint) ? $hint : null, $catMap);
        if ($mapped !== null) {
            return $mapped;
        }
    }
    return null;
}

/**
 * @param list<array> $operations1c
 * @param list<array> $dealsBitrix
 * @return list<array>
 */
function build_sales_unified(array $operations1c, array $dealsBitrix, array $settings): array
{
    $success = $settings['metrics']['bitrix_success_value'] ?? 'Успех';
    $departmentMap = $settings['department_map'] ?? [];
    $rows = [];

    foreach ($operations1c as $record) {
        $sales = isset($record['sales_amount']) ? (float) $record['sales_amount'] : null;
        $resolved = resolve_agent_1c($record['agent'] ?? null, $settings, $record['department'] ?? null);
        $team = $resolved['team'];
        $agentTeams = $resolved['teams'] ?? [$team];
        // Если агент неизвестен — пробуем department_map
        if (str_starts_with((string) $resolved['agent_key'], 'unknown:')) {
            $dept = clean_str($record['department'] ?? null);
            if ($dept !== null) {
                foreach ($departmentMap as $key => $display) {
                    if ($display === $dept || $key === $dept) {
                        $team = $key;
                        break;
                    }
                }
                if ($team === $resolved['team']) {
                    $team = $dept;
                }
                $agentTeams = [$team];
            }
        }
        $rawCategory = clean_str($record['category'] ?? null);
        $relatedService = clean_str($record['related_service_type'] ?? null);
        $category = normalize_category($relatedService ?? $rawCategory, $settings);
        $requestType = resolve_request_type(null, [$rawCategory, $relatedService, $category], $settings);
        $rows[] = [
            'source' => '1c',
            'date' => $record['date_operation'] ?? null,
            'agent_key' => $resolved['agent_key'],
            'agent_display' => $resolved['name_display'],
            'agent_team' => $team ?: 'Без команды',
            'agent_teams' => $agentTeams,
            'agent_is_active' => $resolved['is_active'],
            'agent_raw' => clean_str($record['agent'] ?? null),
            'client' => clean_str($record['client'] ?? null),
            'client_id' => clean_str($record['id_crm'] ?? null),
            'partner_or_supplier' => clean_str($record['supplier'] ?? null),
            'category' => $category,
            'channel' => clean_str($record['channel'] ?? null),
            'card_type' => clean_str($record['card_type'] ?? null),
            'client_type' => clean_str($record['client_type'] ?? null) ?? '1С (нет типа клиента)',
            'request_type' => $requestType,
            'sales_amount' => $sales,
            'profit_ex_vat' => isset($record['profit_ex_vat']) ? to_float($record['profit_ex_vat']) : null,
            'service_fee' => isset($record['service_fee']) ? to_float($record['service_fee']) : null,
            'is_refund' => $sales !== null && $sales < 0,
            'raw_id' => clean_str($record['order_no'] ?? null),
            'date_fallback_used' => null,
        ];
    }

    foreach ($dealsBitrix as $record) {
        $result = clean_str($record['deal_result'] ?? null);
        $isRefund = is_bitrix_refund_deal($record, $settings);
        if ($result !== $success && !$isRefund) {
            continue;
        }
        $sales = isset($record['sales_amount']) ? to_float($record['sales_amount']) : null;
        // Возвраты CRM хранят сумму положительной → в unified как у 1С (минус).
        if ($isRefund && $sales !== null) {
            $sales = -abs($sales);
        }
        $resolved = resolve_agent_bitrix($record['responsible_person'] ?? null, $settings);
        $bTeams = $resolved['teams'] ?? [$resolved['team']];
        $dealNo = $record['deal_no'] ?? null;
        $rawId = null;
        if ($dealNo !== null && $dealNo !== '') {
            $rawId = is_numeric($dealNo) ? (string) (int) $dealNo : (string) $dealNo;
        }
        $rows[] = [
            'source' => 'bitrix',
            'date' => $record['date_for_sales'] ?? null,
            'agent_key' => $resolved['agent_key'],
            'agent_display' => $resolved['name_display'],
            'agent_team' => $resolved['team'] ?: 'Без команды',
            'agent_teams' => $bTeams,
            'agent_is_active' => $resolved['is_active'],
            'agent_raw' => clean_str($record['responsible_person'] ?? null),
            'client' => clean_str($record['client'] ?? null),
            'client_id' => clean_str($record['id_client'] ?? null),
            'partner_or_supplier' => clean_str($record['partner'] ?? null),
            'category' => normalize_category(clean_str($record['category'] ?? null), $settings),
            'channel' => clean_str($record['channel'] ?? null),
            'card_type' => clean_str($record['card_type'] ?? null),
            'client_type' => clean_str($record['client_type'] ?? null),
            'request_type' => resolve_request_type(
                clean_str($record['request_type'] ?? null),
                [
                    clean_str($record['category'] ?? null),
                    normalize_category(clean_str($record['category'] ?? null), $settings),
                ],
                $settings
            ),
            'sales_amount' => $sales,
            'profit_ex_vat' => isset($record['profit_ex_vat']) ? to_float($record['profit_ex_vat']) : null,
            'service_fee' => isset($record['service_fee']) ? to_float($record['service_fee']) : null,
            'is_refund' => $isRefund,
            'raw_id' => $rawId,
            'date_fallback_used' => $record['date_fallback_used'] ?? null,
        ];
    }

    return $rows;
}

/**
 * Справочник агентов для фильтров (из settings + неизвестные из продаж).
 * @return list<array>
 */
function build_agents_dim(array $settings, array $salesUnified): array
{
    $dim = [];
    foreach ($settings['agents'] ?? [] as $agent) {
        $dim[] = [
            'agent_key' => $agent['agent_key'],
            'name_display' => $agent['name_display'] ?? $agent['agent_key'],
            'team' => agent_primary_team($agent),
            'teams' => agent_teams($agent),
            'is_active' => (bool) ($agent['is_active'] ?? true),
            'is_unknown' => 0,
        ];
    }
    $known = array_column($dim, 'agent_key');
    $seen = [];
    foreach ($salesUnified as $row) {
        $key = $row['agent_key'] ?? '';
        if ($key === '' || in_array($key, $known, true) || isset($seen[$key])) {
            continue;
        }
        if (!str_starts_with($key, 'unknown:')) {
            continue;
        }
        $seen[$key] = true;
        $dim[] = [
            'agent_key' => $key,
            'name_display' => $row['agent_display'] ?? $key,
            'team' => $row['agent_team'] ?? 'Без команды',
            'teams' => $row['agent_teams'] ?? [($row['agent_team'] ?? 'Без команды')],
            'is_active' => false,
            'is_unknown' => 1,
        ];
    }
    return $dim;
}
